### ISSUE TITLE
`pack build` fails to export image if extension run image not previously pulled

### ISSUE THREAD

**Issue by cormacpayne** [buildpacks/pack#1804]
### Description
When using `pack build` to create an application image using a builder containing various platform buildpacks and their corresponding build and run extensions, it seems that if the new run image base (from the run extension) wasn't previously pulled down on the machine, then the application image cannot be exported.

Below are the verbose logs from the export phase where the new run image base _has not_ been pulled down onto the machine:

```
===> EXPORTING
Running the 'exporter' on OS 'linux' with:
Container Settings:
Args: '/cnb/lifecycle/exporter -daemon -launch-cache /launch-cache -log-level debug -app /workspace -cache-dir /cache -stack /layers/stack.toml <registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615132919112539'
System Envs: 'CNB_USER_ID=1000 CNB_GROUP_ID=1000 CNB_PLATFORM_API=0.10'
Image: 'buildpacksio/lifecycle:0.15.2'
User: 'root'
Labels: 'map[author:pack]'
Host Settings:
Binds: 'pack-cache-cbl-mariner-aca_run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615132919112539-f07c32078b84.build:/cache /var/run/docker.sock:/var/run/docker.sock pack-cache-cbl-mariner-aca_run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615132919112539-f07c32078b84.launch:/launch-cache pack-layers-jjteiimkuq:/layers pack-app-vfbuhpafuu:/workspace'
Network Mode: ''
[exporter] Using runImage from analyzed metadata 'mcr.microsoft.com/dotnet/aspnet:7.0.5-cbl-mariner2.0'
[exporter] ERROR: failed to export: get run image top layer SHA: image "<registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615132919112539" has no layers
```

If I then use `docker pull` to pull down the new run image specified in the extension, the application image is exported successfully:

```
===> EXPORTING
Running the 'exporter' on OS 'linux' with:
Container Settings:
Args: '/cnb/lifecycle/exporter -daemon -launch-cache /launch-cache -log-level debug -app /workspace -cache-dir /cache -stack /layers/stack.toml <registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615133552173374'
System Envs: 'CNB_USER_ID=1000 CNB_GROUP_ID=1000 CNB_PLATFORM_API=0.10'
Image: 'buildpacksio/lifecycle:0.15.2'
User: 'root'
Labels: 'map[author:pack]'
Host Settings:
Binds: 'pack-cache-cbl-mariner-aca_run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615133552173374-0a165b2ebdc2.build:/cache /var/run/docker.sock:/var/run/docker.sock pack-cache-cbl-mariner-aca_run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615133552173374-0a165b2ebdc2.launch:/launch-cache pack-layers-lpvxlnglpr:/layers pack-app-iexgxlzvee:/workspace'
Network Mode: ''
[exporter] Using runImage from analyzed metadata 'mcr.microsoft.com/dotnet/aspnet:7.0.5-cbl-mariner2.0'
[exporter] Reading buildpack directory: /layers/capps_dotnet-7.0
[exporter] Reading buildpack directory item: launch.toml
[exporter] Reading buildpack directory item: run-script
[exporter] Reading buildpack directory item: run-script.toml
[exporter] Processing buildpack directory: /layers/capps_dotnet-7.0
[exporter] Processing launch layer: /layers/capps_dotnet-7.0/run-script
[exporter] Reusing tarball for layer "capps/dotnet-7.0:run-script" with SHA: sha256:26093b5f94f9c6a4314497485f643f2bb72363d32991d9b3176be5cc3887a2de
[exporter] Adding layer 'capps/dotnet-7.0:run-script'
[exporter] Layer 'capps/dotnet-7.0:run-script' SHA: sha256:26093b5f94f9c6a4314497485f643f2bb72363d32991d9b3176be5cc3887a2de
[exporter] Found SBOM of type launch for at /layers/sbom/launch
[exporter] Reusing tarball for layer "launch.sbom" with SHA: sha256:36f9de0748c420980a265697f38f76b0f9ae9b7862d3467b76b638706e637425
[exporter] Adding layer 'launch.sbom'
[exporter] Layer 'launch.sbom' SHA: sha256:36f9de0748c420980a265697f38f76b0f9ae9b7862d3467b76b638706e637425
[exporter] Layer 'slice-1' SHA: sha256:0babcc029d5f7df191832a5db4c2eb34548eabdd058a9ac31b378117331d16d2
[exporter] Adding 1/1 app layer(s)
[exporter] Reusing tarball for layer "launcher" with SHA: sha256:6c749df2cfd63eaf8d564c60026df199deccb92225c279b7e858e7cae4dfbe14
[exporter] Adding layer 'launcher'
[exporter] Layer 'launcher' SHA: sha256:6c749df2cfd63eaf8d564c60026df199deccb92225c279b7e858e7cae4dfbe14
[exporter] Reusing tarball for layer "config" with SHA: sha256:7987d808359d7236349aae784ec2f913f451d30a1241fcbca454392c6e8d5bde
[exporter] Adding layer 'config'
[exporter] Layer 'config' SHA: sha256:7987d808359d7236349aae784ec2f913f451d30a1241fcbca454392c6e8d5bde
[exporter] Reusing tarball for layer "process-types" with SHA: sha256:e7adcf8438e4e7fd9efd6b7ec714bc7031e2fba9bcf42d5450a5c88957f2f391
[exporter] Adding layer 'process-types'
[exporter] Layer 'process-types' SHA: sha256:e7adcf8438e4e7fd9efd6b7ec714bc7031e2fba9bcf42d5450a5c88957f2f391
[exporter] Adding label 'io.buildpacks.lifecycle.metadata'
[exporter] Adding label 'io.buildpacks.build.metadata'
[exporter] Adding label 'io.buildpacks.project.metadata'
[exporter] Setting CNB_LAYERS_DIR=/layers
[exporter] Setting CNB_APP_DIR=/workspace
[exporter] Setting CNB_PLATFORM_API=0.10
[exporter] Setting CNB_DEPRECATION_MODE=quiet
[exporter] Prepending /cnb/process and /cnb/lifecycle to PATH
[exporter] Setting WORKDIR: '/workspace'
[exporter] Setting default process type 'run-script'
[exporter] Setting ENTRYPOINT: '/cnb/process/run-script'
[exporter] Saving <registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615133552173374...
[exporter] *** Images (d689d05de34f):
[exporter]       <registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615133552173374
[exporter]
[exporter] *** Image ID: d689d05de34f7ec9bbcffff5b9d13020307874927364159de97d08368cb3af24
[exporter] Reading buildpack directory: /layers/capps_dotnet-7.0
[exporter] Reading buildpack directory item: launch.toml
[exporter] Reading buildpack directory item: run-script
[exporter] Reading buildpack directory item: run-script.toml
```

### Proposed solution
Using the same daemon that can pull down the `buildpacksio/lifecycle`, builder and stack images if not present on the machine, the new run image defined in the extension would also be pulled down if not present.

It appears that the exporter already has context about what this new run image is via the logs above:

```
[exporter] Using runImage from analyzed metadata 'mcr.microsoft.com/dotnet/aspnet:7.0.5-cbl-mariner2.0'
```

So I was hoping the daemon could check if this image is present locally first before proceeding.

### Describe alternatives you've considered
We could manually pull down the run image before running `pack build`, but we'd have to ensure that all possible run images bundled as extensions within the builder are present since we won't know which platform a user is targeting when using the builder until the buildpacks decide.

### Additional context
- [ ] This feature should be documented somewhere

**Comment by cormacpayne** [comment]
@natalieparellano shoot, not sure how I missed the existing issue (GitHub search filters strike again 😢) -- I just updated to the `pack` version you mentioned and the latest `buildpacksio/lifecycle` available (`0.17.0-rc.1`) and am unfortunately still seeing this error, with slightly different verbiage prior to it:

```
===> EXPORTING
Running the 'exporter' on OS 'linux' with:
Container Settings:
Args: '/cnb/lifecycle/exporter -daemon -launch-cache /launch-cache -log-level debug -app /workspace -cache-dir /cache -stack /layers/stack.toml <registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615142620621473'
System Envs: 'CNB_USER_ID=1000 CNB_GROUP_ID=1000 CNB_PLATFORM_API=0.10'
Image: 'index.docker.io/buildpacksio/lifecycle:0.17.0-rc.1'
User: 'root'
Labels: 'map[author:pack]'
Host Settings:
Binds: 'pack-cache-cbl-mariner-aca_run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615142620621473-230f7014d610.build:/cache /var/run/docker.sock:/var/run/docker.sock pack-cache-cbl-mariner-aca_run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615142620621473-230f7014d610.launch:/launch-cache pack-layers-lkuzqarmrk:/layers pack-app-nxwcafucvm:/workspace'

Network Mode: ''
[exporter] Starting exporter...
[exporter] Parsing inputs...
[exporter] Ensuring privileges...
[exporter] Executing command...
[exporter] ERROR: failed to export: get run image top layer SHA: image "<registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615142620621473" has no layers
```

I'm happy to provide any additional information or try any other versions of the tools involved!

**Comment by natalieparellano** [comment]
Ah, apologies @cormacpayne - I may have steered you wrong. It appears that pack `0.30.0-pre2` may be broken with lifecycle `0.17.0-rc.1`. 

I suspect the output from above may still be from an older version of pack, because I see `System Envs: 'CNB_USER_ID=1000 CNB_GROUP_ID=1000 CNB_PLATFORM_API=0.10'` - newer pack would try to use `CNB_PLATFORM_API=0.12`. I was trying to repro this myself but instead I got this error:

```
===> ANALYZING
Running the analyzer on OS linux with:
Container Settings:
  Args: /cnb/lifecycle/analyzer -log-level debug -daemon -stack /layers/stack.toml -run-image cnbs/sample-stack-run:alpine -launch-cache /launch-cache hello-extensions
  System Envs: CNB_USER_ID=1000 CNB_GROUP_ID=1001 CNB_PLATFORM_API=0.12
  Image: buildpacksio/lifecycle:0.17.0-rc.1
  User: root
  Labels: map[author:pack]
Host Settings:
  Binds: /var/run/docker.sock:/var/run/docker.sock pack-cache-library_hello-extensions_latest-605c1c945967.launch:/launch-cache pack-layers-rnnzoquprm:/layers pack-app-rgjtjvruhv:/workspace
  Network Mode: host
[analyzer] flag provided but not defined: -stack
...
ERROR: failed to build: executing lifecycle: failed with status code: 2
```

This particular bug will be fixed in https://github.com/buildpacks/pack/pull/1789 and together with a couple of other fixes (cc @jjbustamante) we hope to cut a new pre-release in the next few days.

In the meantime, I tried to repro the above using lifecycle `0.17.0-pre.2`, and I did see the image being pulled:

* Using the `runext/update-samples` branch of the sample repo
* Updated `builders/alpine/builder.toml` to contain:
```
[lifecycle]
version = "0.17.0-pre.2"
```
* (before the build)`$  docker image ls` has:
```
REPOSITORY               TAG           IMAGE ID       CREATED        SIZE
registry                 2             4bb5ea59f8e0   33 hours ago   24MB
cnbs/sample-stack-run    alpine        76c07164e3cb   6 months ago   8.11MB
buildpacksio/lifecycle   0.17.0-rc.1   77123df190b0   43 years ago   32.5MB
```
* (note `localhost:5000/run-image-curl` is NOT present)
* `$  pack build hello-extensions   --builder localhost:5000/extensions-builder   --env BP_EXT_DEMO=1   --env BP_REQUIRES=curl   --path $PWD/apps/java-maven   --pull-policy always   --network host   --verbose` (`BP_REQUIRES=curl` invokes the extension that does run image switching)
* The output has:
```
[detector] Found a run.Dockerfile from extension 'samples/curl' setting run image to 'localhost:5000/run-image-curl' 
...
===> BUILDING
...
Pulling image localhost:5000/run-image-curl
latest: Pulling from run-image-curl
...
Status: Downloaded newer image for localhost:5000/run-image-curl:latest
...
===> EXPORTING
Running the exporter on OS linux with:
Container Settings:
  Args: /cnb/lifecycle/exporter -daemon -launch-cache /launch-cache -log-level debug -app /workspace -cache-dir /cache -run /layers/run.toml hello-extensions
  System Envs: CNB_USER_ID=1000 CNB_GROUP_ID=1001 CNB_PLATFORM_API=0.12
...
Successfully built image hello-extensions
```
* (note `CNB_PLATFORM_API=0.12`)
* (after the build)`$  docker image ls` has:
```
REPOSITORY                          TAG            IMAGE ID       CREATED          SIZE
localhost:5000/run-image-curl       latest         b674d18e2cf7   12 minutes ago   22.7MB
registry                            2              4bb5ea59f8e0   33 hours ago     24MB
cnbs/sample-stack-run               alpine         76c07164e3cb   6 months ago     8.11MB
localhost:5000/extensions-builder   latest         927794559adb   43 years ago     55.1MB
buildpacksio/lifecycle              0.17.0-pre.2   16ea1bbc6247   43 years ago     32.3MB
hello-extensions                    latest         b65cc3ab2b77   43 years ago     25.1MB
buildpacksio/lifecycle              0.17.0-rc.1    77123df190b0   43 years ago     32.5MB
```
* (note `localhost:5000/run-image-curl` IS present)

Is it possible to check if you have `Pulling image <registry>/cbl-mariner-aca:run-dotnet-aspnet-7.0.7-cbl-mariner2.0-20230615142620621473` in your logs? There are a couple of places where this could go wrong, but barring the pack version being older it's possible we don't have the right creds to pull the run image.

**Comment by natalieparellano** [comment]
@cormacpayne - that's very interesting. If you run `pack builder inspect` on your builder, what Platform API versions do you see under `Lifecycle`, e.g.:
```
Lifecycle:
  Version: 0.15.2
...
  Platform APIs:
    Deprecated: (none)
    Supported: 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 0.10
```
Perhaps the `CNB_PLATFORM_API` though is a red herring - I'll grab a Windows environment to see if I could reproduce your issue there, it might have to do with the filepaths that are generated when copying files out of a container. We grab `<layers>/analyzed.toml` out of the `detect` container to determine if the run image changed.

**Comment by cormacpayne** [comment]
@natalieparellano ah ha, I found the culprit thanks to the `pack builder inspect` command -- the `Created By` field told me the builder was created via a `0.29.0` version of `pack`, so it looks like I _was_ using the correct version of the CLI, but I just hadn't refreshed my builder to be created by it.

I just ran through this scenario with the updated builder and we're all good now -- the extension run image is being pulled down after a wipe of the Docker images and cache!

Sorry for the false alarm, and thanks again for all of the help! 😃
